File: C/chapitre5/Exercice11.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

pthread_mutex_t m_lect = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t m_ecrit = PTHREAD_MUTEX_INITIALIZER;
int lecteurs = 0;
char db[15];

void *lecteur(void *arg) {
    while (1) {
        pthread_mutex_lock(&m_lect);
        lecteurs++;
        if (lecteurs == 1) {
            pthread_mutex_lock(&m_ecrit);
        }
        pthread_mutex_unlock(&m_lect);
        int idx = rand() % 15;
        printf("Lit [%d]=%d\n", idx, db[idx]);
        sleep(1 + rand() % 3);
        pthread_mutex_lock(&m_lect);
        lecteurs--;
        if (lecteurs == 0) {
            pthread_mutex_unlock(&m_ecrit);
        }
        pthread_mutex_unlock(&m_lect);
        sleep(1 + rand() % 3);
    }
    return NULL;
}

void *redacteur(void *arg) {
    while (1) {
        pthread_mutex_lock(&m_ecrit);
        int idx = rand() % 15;
        db[idx] = rand() % 100;
        printf("Ecrit [%d]=%d\n", idx, db[idx]);
        pthread_mutex_unlock(&m_ecrit);
        sleep(1 + rand() % 10);
    }
    return NULL;
}

int main(void) {
    int L;
    int R;
    scanf("%d%d", &L, &R);
    srand(time(NULL));
    pthread_t t[L + R];
    int i;
    for (i = 0; i < L; i++) {
        pthread_create(&t[i], NULL, lecteur, NULL);
    }
    for (i = 0; i < R; i++) {
        pthread_create(&t[L + i], NULL, redacteur, NULL);
    }
    for (i = 0; i < L + R; i++) {
        pthread_join(t[i], NULL);
    }
    return 0;
}

File: Python/chapitre5/Exercice11.py
import threading
import time
import random

m_lect = threading.Lock()
m_ecrit = threading.Lock()
lecteurs = 0
db = [0] * 15

def lecteur():
    global lecteurs
    while True:
        m_lect.acquire()
        lecteurs += 1
        if lecteurs == 1:
            m_ecrit.acquire()
        m_lect.release()
        idx = random.randint(0, 14)
        print("Lit [", idx, "]=", db[idx])
        m_lect.acquire()
        lecteurs -= 1
        if lecteurs == 0:
            m_ecrit.release()
        m_lect.release()
        time.sleep(random.randint(1, 3))

def redacteur():
    while True:
        m_ecrit.acquire()
        idx = random.randint(0, 14)
        db[idx] = random.randint(0, 100)
        print("Ecrit [", idx, "]=", db[idx])
        m_ecrit.release()
        time.sleep(random.randint(1, 10))

if __name__ == "__main__":
    L = int(input("Nombre lecteurs: "))
    R = int(input("Nombre rédacteurs: "))
    threads = []
    for i in range(L):
        t = threading.Thread(target=lecteur)
        t.start()
        threads.append(t)
    for i in range(R):
        t = threading.Thread(target=redacteur)
        t.start()
        threads.append(t)
    for t in threads:
        t.join()